File: backtrack/array_sum_combinations/array_sum_combinations.jl
function product(head:: Array{Int, 1})
    [[x] for x in head]
end

function product(head::Array{Int, 1}, tail::Array{Int, 1}...)
    subproducts = product(tail...)
    [[x, p...] for p in subproducts for x in head]
end

function Counter(iterable)
    hmap = Dict{Int, Int}()
    for el in iterable
        hmap[el] = get(hmap, el, 0) + 1
    end
    hmap
end

function sum_combinations(target, arrs...)
    counter_last = Counter(arrs[end])
    ans = []
    for combination in product(arrs...)
        difference = target - sum(combination)
        if difference in keys(counter_last)
            for _ in 0:counter_last[difference]
                push!(ans, [combination ; difference])
            end
        end
    end
    ans
end


A = [1, 2, 3, 3]
B = [2, 3, 3, 4]
C = [1, 2, 2, 2]
target = 7

println(sum_combinations(target, A, B, C))

File: backtrack/array_sum_combinations/array_sum_combinations.js
"use strict"

// Sigh JS ...

// cute two line generator although it reverses order
function* product(head, ...tail) {
    const remaining = tail.length > 0 ? product(...tail) : [[]]
    for (const r of remaining) for (const h of head) yield [h, ...r]
}

function Counter(iterable) {
    const hmap = new Map()
    for (const el of iterable) {
        hmap.set(el, (hmap.get(el) || 0) + 1)
    }
    return hmap
}

const sum = iterable => iterable.reduce((pv, cv) => pv + cv)

function* sumCombinations(target, ...arrs) {
    const last = arrs[arrs.length - 1]
    const rest = arrs.slice(0, -1)
    const counterLast = new Counter(last)
    for (const combination of product(...rest)) {
        const difference = target - sum(combination)
        if (counterLast.has(difference)) {
            const times = counterLast.get(difference).length
            for (let i=0; i< times; i++){
              yield combination.concat(difference)
            }
        }
    }
}

const A = [1, 2, 3, 3]
const B = [2, 3, 3, 4]
const C = [1, 2, 2, 2]
const target = 7

for (const combination of sumCombinations(target, A, B, C)) {
    console.log(combination)
}
